engineapi/proxy.go

```diff
@@ -19,77 +23,296 @@ import (
+		for instanceManagerImage, handle := range imImageHandled {
+			im, err := h.ds.GetInstanceManagerRO(handle.instanceManagerName)
+			if err != nil {
+				h.logger.WithError(err).Warnf("Failed to get %v instance manager for proxy", handle.instanceManagerName)
+				continue
+			}
+
+			log := getLoggerForProxyHandler(h.logger, im)
+
+			for clientIndex, client := range handle.Clients {
+				if !client.ShouldResync(h.clock) {
+					continue
+				}
+
+				cLog := client.GetLogger()
+
+				if client.ShouldDelete() {
+					cLog.Debug("Deleting proxy client")
+					client.SetNextGC(h.clock)
+
+					newIndex, err := h.getNewClientIndex(handle.Clients)
+					if err != nil {
+						cLog.WithError(err).Debug("Cannot replace proxy client")
+					} else {
+						cLog.Debug("Replacing with new proxy client")
+						h.NewHandleClient(im, newIndex)
+					}
+
+					client.SetNextResync(h.clock, h.resyncPeriod)
+					break
+				}
+
+				h.gc(instanceManagerImage, clientIndex)
+
+				if client.ShouldCreate(im) {
+					cLog.Debug("Creating proxy client")
+
+					if err := client.Start(clientIndex, im, h.logger, h.ds, h.stopCh); err != nil {
+						cLog.Warn("Failed to start proxy client")
+					}
+				}
+
+				client.SetNextResync(h.clock, h.resyncPeriod)
+			}
+
+			exceedMaxConnection := true
+			for _, client := range handle.Clients {
+				if client.GetInUse() >= MaxProxyClientConcurrentInUse {
+					continue
+				}
+
+				exceedMaxConnection = false
+				break
+			}
+			if len(handle.Clients) != 0 && exceedMaxConnection {
+				log.Debugf("Exceed proxy client max concurrent use limit:%v", MaxProxyClientConcurrentInUse)
+
+				newIndex, err := h.getNewClientIndex(handle.Clients)
+				if err != nil {
+					log.WithError(err).Debugf("Cannot add more proxy client")
+				} else {
+					log.Debugf("Add proxy client-%v to handle concurrent use", newIndex)
+					h.NewHandleClient(im, newIndex)
+				}
+			}
+		}
+	}
+}
+
+func (h *EngineClientProxyHandler) gc(instanceManagerImage string, clientIndex int) {
+	client, exist := h.GetHandleClient(instanceManagerImage, clientIndex)
+	if !exist {
+		h.logger.Debugf("Proxy %v client-%v already gone", instanceManagerImage, clientIndex)
+		return
+	}
+
+	log := client.GetLogger()
+
+	nextGC := client.GetNextGC()
+	now := h.clock.Now()
+	if nextGC == nil || now.Before(*nextGC) {
+		return
+	}
+
+	msg := ""
+	if client.GetInUse() != 0 {
+		msg = fmt.Sprintf("Cannot stop proxy client with %v still in use", client.GetInUse())
+		if client.GetGCRetry() < ProxyGCMaxRetry {
+			log.Warnf(msg)
+			client.SetNextGC(h.clock)
+			return
+		}
+	}
+
+	log.Debug("Stopping proxy client")
+	if err := client.Stop(); err != nil {
+		msg = "Failed to stop proxy client"
+		if client.GetGCRetry() < ProxyGCMaxRetry {
+			log.WithError(err).Warnf(msg)
+			client.SetNextGC(h.clock)
+			return
+		}
+	}
+
+	if msg != "" {
+		log.Warn("Exceed max retry limit, force delete proxy client")
+	}
+
+	h.lock.Lock()
```

Comment:
I think this lock should be for the whole scope of this method instead of only protecting the map.

Reply:
We have Rlock protecting others in the method.